Download the TF-DNA binding gcPBM data for Max, Myc, and Mad from Blackboard: ‘Max.txt’
for TF Max, ‘Mad.txt’ for TF Mad, and ‘Myc.txt’ for TF Myc. Each file contains two columns,
where the first column represents the binding sequence and the second column shows the
binding affinity. Utilize pandas to read these files into a DataFrame format. Once the data is
loaded, count and report the number of sequences for each TF. [0.5pt]

In [1]:
import pandas as pd

In [2]:
mad_df = pd.read_table('Mad.txt', delimiter='\t', header = None, names = ['binding sequence', 'binding affinity'])
mad_df 

binding sequence  binding affinity
0     GGGCATGAAAGCCACCTCGTGGTTTTGCAGCAAGTG          8.763428
1     CGGGGAGGAGCCACCCACGCGCCACAGCAAGTCCTG          7.903412
2     GGCGGCCCGGGCATCCACGTGGGTCGGGCTCCTGGT         10.208874
3     TGCGGCTTCGGCTTCCACGCGGCATCCTAGGTAAGC          8.273974
4     TGCTCGCCTGCCTGCCATGTGCAGCGGCCCCTTTGT          8.325185
...                                    ...               ...
7529  CGGCCCCCGGCCCGCCATGTGGCGCGGCTGACCGCG          8.361708
7530  GTACCCGCGCTGTTGCGCGTGGGTCCCGGGAGCGGC          7.964677
7531  GCGCAGGAGCGCCGACACGCGTCAGCACTAGCCGGC          7.808526
7532  CGAGGCAGGTGCGGTCACGTGACCCGGCGGCGCTGC          8.852808
7533  CTTGAAAACAGAGGCCATGTGGTGGCCTGTTACATA          8.644618

[7534 rows x 2 columns]

The TF Mad datset contains 7534 sequences.

In [3]:
max_df = pd.read_table('Max.txt', delimiter = '\t', header = None, names = ['binding sequence', 'binding affinity'])
max_df

binding sequence  binding affinity
0     GAAGCCCTGGCGGGGCGCGTGCCCGCCGCCGCCGCC          3.737829
1     TCCCGGGGCTAGAGGCATGTGGACTCAGGAGGATGA          3.895754
2     GGCGCTGCCGGACTGCGCGTGGAGTGGCGCGCTGCT          3.797545
3     GCAGCTCGCCCGGGCCACGCGGTGCCGGCCCTCGGC          4.080446
4     GGAATTACCTTCATCCACATGGTTGAAGCTGGCTGG          4.126748
...                                    ...               ...
8563  AGCAAAGCTGGCCCCCACGTTGTCCTTCCGCAGAAC          4.010130
8564  GGGACCCTGCCTTCCCGCGTGGCCGCACCGAGACGA          4.036789
8565  AGAAGAACCCGGGCGCACGTGGCCGACTCCAGGCTG          4.418053
8566  GGGGCCCGGGTGGGGCATGTGGGGACAACGCCTGGA          3.939968
8567  AAGCAAACCCCCCAACACGTGTTTCCTAGTTGGGAA          4.511081

[8568 rows x 2 columns]

The TF Max dataset contains 8568 sequences.

In [ ]:
myc_df = pd.read_table('Myc.txt', delimiter = '\t', header = None, names = ['binding sequence', 'binding affinity'])
myc_df

The TF Myc dataset contains 6926 sequences.

Write a function to convert DNA sequences into a one-hot encoding format. This should be done
for both mononucleotide (1mer) and dinucleotide (2mer) formats. Use the following mappings:
a. For 1mer: A: 1000, C: 0100, G: 0010, T: 0001
b. For 2mer: AA: 1000000000000000, AC: 0100000000000000, AG: 0010000000000000,
AT: 0001000000000000, …, TT: 0000000000000001
For instance, given the input sequence 'AAC', the output should be '100010000100' in 1mer
encoding. In 2mer encoding, the output should be '10000000000000000100000000000000'.
[2pt]

In [5]:
def one_hot_encode_1mer(seq):
    temp = list(seq.replace('A', '1000').replace('C', '0100').replace('G', '0010').replace('T','0001'))
    return [int(x) for x in temp]
print(one_hot_encode_1mer('AAC'))

[1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0]


In [6]:
def one_hot_encode_2mer(seq):
    str = ''
    for i in range(len(seq)-1):
        str += seq[i:i+2].replace('AA', '1000000000000000').replace('AC', '0100000000000000').replace('AG', '0010000000000000').replace('AT', '0001000000000000').replace('CA', '0000100000000000').replace('CC', '0000010000000000').replace('CG', '0000001000000000').replace('CT', '0000000100000000').replace('GA', '0000000010000000').replace('GC', '0000000001000000').replace('GG', '0000000000100000').replace('GT', '0000000000010000').replace('TA', '0000000000001000').replace('TC', '0000000000000100').replace('TG', '0000000000000010').replace('TT', '0000000000000001')
    return [int(x) for x in list(str)]
print(one_hot_encode_2mer('AAC'))

[1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]


Normalize the binding affinity values, which is located in the second column, using the min-max
normalization method. Min-max normalization will scale the binding affinity values to a range
between 0 and 1, based on the minimum and maximum values in the data. [1pt]

In [7]:
def normalize_binding(df):
    min_val = min(df['binding affinity'])
    max_val = max(df['binding affinity'])
    df['binding affinity'] = df['binding affinity'].apply(lambda x: (x - min_val)/(max_val - min_val))
    return df

normalize_binding(mad_df)
normalize_binding(max_df)
normalize_binding(myc_df)

binding sequence  binding affinity
0     ACCGACCGGCGCGGGCACGAGGCAATGGCGGCCGGG          0.174288
1     AACAGCGCCACCGGCCTCGTGCACTTCTTCCACTGT          0.217422
2     GCGGCCGGTCTGCACCATGCTGCGAACGTCCGTCCT          0.133137
3     TTATTGAGCCCCTACCATGTGCCAGGCCCTGGGCTA          0.311328
4     CCCCTCAGCTGCTTCCTCGTGGCGCTGATCATCTGG          0.319377
...                                    ...               ...
6921  GCCCCCTGCGGGAGCCGCGTGGACCGCCCCGGTGGC          0.389523
6922  GGACGTCTCAAAACACACGTGTGACTTAGATGGTAA          0.520336
6923  TCTGCCCAGTCCCTCCACATGGGAAATATAGCACTG          0.294541
6924  AGGACTACAAGTTCCAGCGTGCTTTGCTCGATTTCG          0.058199
6925  CGGGAGGGGCGGGGCCACGCTGCGGGCCCGGGCCAT          0.249490

[6926 rows x 2 columns]

(Cross-validation) Use 10-fold cross validation to train these two models. [1pt]

(Regularization) Based on the characteristics of the data and the nature of the task, choose an
appropriate type of regularization (L1, L2, or Elastic Net). Justify your choice with a brief
explanation. Additionally, describe your strategy to find the optimal lambda value, which is the
regularization parameter that determines the strength of the penalty. Note that you are not
required to write code for this part; instead, focus on explaining the methodology you plan to
use. [1.5pt]

**Elastic net regularization may be appropriate for this task because there is some multicollinearity in the features extracted by the model but there are still numerous relevant features, each of which has a small but non-zero weight. Therefore, we would not want to eliminate many features in favor of a few features like in L1 regularization. We also would not like to include as many features as possible (as in L2 regularization) because there are some features that are collinear and which measure similar properties. For instance, if the model is able to learn that sequences of alternating A and T have higher binding affinity scores, then we could capture that in a few features. We wouldn't need a feature that scans for "ATATATAT" and another feature that scans for "ATATATATAT", and even more features for even longer AT repeats. Therefore, we would want to prune those collinear features using Elastic Net regularization to zero their weights.**

**Given that we will use Elastic Net, we can obtain an optimal labmda parameter value via many approaches which fall under the umbrella of hyperparameter optimization. One simple approach is that we can conduct gradient descent with respect to lambda in order to minimize the loss function. Alternatively, we can do a random search by sampling random values in the hyperparameter space. However, cursory online research shows that common practice is to do grid search, which is similar to randoms search but the search space is bounded by user-selected range values. Therefore, we are instructing the algorithm to search within a range of reasonable values and find the set of values which minimize/maximize some loss or performance metric. In this case, I would opt for grid search because it achieves relatively optimal lambda values while also being quick and computationally efficient.**

In [ ]:
import numpy as np
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import KFold
from sklearn.linear_model import LinearRegression, Lasso, Ridge, ElasticNet
from sklearn.metrics import mean_squared_error

In [8]:
max_sequences = max_df['binding sequence'].tolist()
max_values = max_df['binding affinity'].tolist()

# One-hot encode the sequences
max_X = np.array([one_hot_encode_1mer(seq) for seq in max_sequences])
max_y = np.array(max_df['binding affinity'])

max_cv = KFold(n_splits=10, shuffle=True, random_state=1)

In [9]:
# results
mse_scores = []; r_squared_scores = []

# Initialize Linear Regression model
#model = LinearRegression()
#model = Lasso(alpha=0.001)
#model = Ridge(alpha=0.001)
model = ElasticNet(alpha=0.001, l1_ratio=0.5)

# Loop over each fold
for train_index, valid_index in max_cv.split(max_X):
    # Split data
    X_train, X_valid = max_X[train_index], max_X[valid_index]
    y_train, y_valid = max_y[train_index], max_y[valid_index]

    # Fit the model
    model.fit(X_train, y_train)
    
    # Predict on valid set
    predictions = model.predict(X_valid)
    
    # Evaluate the model
    mse = mean_squared_error(y_valid, predictions)
    mse_scores.append(mse)
    
    r_squared = model.score(X_valid, y_valid)
    r_squared_scores.append(r_squared)
    
average_mse = np.mean(mse_scores)
average_r_squared = np.mean(r_squared_scores)

print(f'TRIAL FOR MAX TF 1-MER: Average MSE: {average_mse} Average r_squared: {average_r_squared}')

TRIAL FOR MAX TF 1-MER: Average MSE: 0.006311913071929501 Average r_squared: 0.78088946154413


In [10]:
max_sequences = max_df['binding sequence'].tolist()
max_values = max_df['binding affinity'].tolist()

# One-hot encode the sequences
max_X = np.array([one_hot_encode_2mer(seq) for seq in max_sequences])
max_y = np.array(max_df['binding affinity'])

max_cv = KFold(n_splits=10, shuffle=True, random_state=1)

In [11]:
# results
mse_scores = []; r_squared_scores = []

# Initialize Linear Regression model
#model = LinearRegression()
#model = Lasso(alpha=0.001)
#model = Ridge(alpha=0.001)
model = ElasticNet(alpha=0.001, l1_ratio=0.5)

# Loop over each fold
for train_index, valid_index in max_cv.split(max_X):
    # Split data
    X_train, X_valid = max_X[train_index], max_X[valid_index]
    y_train, y_valid = max_y[train_index], max_y[valid_index]

    # Fit the model
    model.fit(X_train, y_train)
    
    # Predict on valid set
    predictions = model.predict(X_valid)
    
    # Evaluate the model
    mse = mean_squared_error(y_valid, predictions)
    mse_scores.append(mse)
    
    r_squared = model.score(X_valid, y_valid)
    r_squared_scores.append(r_squared)
    
average_mse = np.mean(mse_scores)
average_r_squared = np.mean(r_squared_scores)

print(f'TRIAL FOR MAX TF 2-MER: Average MSE: {average_mse} Average r_squared: {average_r_squared}')

TRIAL FOR MAX TF 2-MER: Average MSE: 0.004261243948972678 Average r_squared: 0.8520113133526607


In [12]:
mad_sequences = mad_df['binding sequence'].tolist()
mad_values = mad_df['binding affinity'].tolist()

# One-hot encode the sequences
mad_X = np.array([one_hot_encode_1mer(seq) for seq in mad_sequences])
mad_y = np.array(mad_df['binding affinity'])

mad_cv = KFold(n_splits=10, shuffle=True, random_state=1)

In [13]:
# results
mse_scores = []; r_squared_scores = []

# Initialize Linear Regression model
#model = LinearRegression()
#model = Lasso(alpha=0.001)
#model = Ridge(alpha=0.001)
model = ElasticNet(alpha=0.001, l1_ratio=0.5)

# Loop over each fold
for train_index, valid_index in max_cv.split(mad_X):
    # Split data
    X_train, X_valid = mad_X[train_index], mad_X[valid_index]
    y_train, y_valid = mad_y[train_index], mad_y[valid_index]

    # Fit the model
    model.fit(X_train, y_train)
    
    # Predict on valid set
    predictions = model.predict(X_valid)
    
    # Evaluate the model
    mse = mean_squared_error(y_valid, predictions)
    mse_scores.append(mse)
    
    r_squared = model.score(X_valid, y_valid)
    r_squared_scores.append(r_squared)
    
average_mse = np.mean(mse_scores)
average_r_squared = np.mean(r_squared_scores)

print(f'TRIAL FOR MAD TF 1-MER: Average MSE: {average_mse} Average r_squared: {average_r_squared}')

TRIAL FOR MAD TF 1-MER: Average MSE: 0.010255679123879247 Average r_squared: 0.7711734757652452


In [14]:
mad_sequences = mad_df['binding sequence'].tolist()
mad_values = mad_df['binding affinity'].tolist()

# One-hot encode the sequences
mad_X = np.array([one_hot_encode_2mer(seq) for seq in mad_sequences])
mad_y = np.array(mad_df['binding affinity'])

mad_cv = KFold(n_splits=10, shuffle=True, random_state=1)

In [15]:
# results
mse_scores = []; r_squared_scores = []

# Initialize Linear Regression model
#model = LinearRegression()
#model = Lasso(alpha=0.001)
#model = Ridge(alpha=0.001)
model = ElasticNet(alpha=0.001, l1_ratio=0.5)

# Loop over each fold
for train_index, valid_index in max_cv.split(mad_X):
    # Split data
    X_train, X_valid = mad_X[train_index], mad_X[valid_index]
    y_train, y_valid = mad_y[train_index], mad_y[valid_index]

    # Fit the model
    model.fit(X_train, y_train)
    
    # Predict on valid set
    predictions = model.predict(X_valid)
    
    # Evaluate the model
    mse = mean_squared_error(y_valid, predictions)
    mse_scores.append(mse)
    
    r_squared = model.score(X_valid, y_valid)
    r_squared_scores.append(r_squared)
    
average_mse = np.mean(mse_scores)
average_r_squared = np.mean(r_squared_scores)

print(f'TRIAL FOR MAD TF 2-MER: Average MSE: {average_mse} Average r_squared: {average_r_squared}')

TRIAL FOR MAD TF 2-MER: Average MSE: 0.0067631008309956735 Average r_squared: 0.8491122657585828


In [16]:
myc_sequences = myc_df['binding sequence'].tolist()
myc_values = myc_df['binding affinity'].tolist()

# One-hot encode the sequences
myc_X = np.array([one_hot_encode_1mer(seq) for seq in myc_sequences])
myc_y = np.array(myc_df['binding affinity'])

myc_cv = KFold(n_splits=10, shuffle=True, random_state=1)

In [17]:
# results
mse_scores = []; r_squared_scores = []

# Initialize Linear Regression model
#model = LinearRegression()
#model = Lasso(alpha=0.001)
#model = Ridge(alpha=0.001)
model = ElasticNet(alpha=0.001, l1_ratio=0.5)

# Loop over each fold
for train_index, valid_index in max_cv.split(myc_X):
    # Split data
    X_train, X_valid = myc_X[train_index], myc_X[valid_index]
    y_train, y_valid = myc_y[train_index], myc_y[valid_index]

    # Fit the model
    model.fit(X_train, y_train)
    
    # Predict on valid set
    predictions = model.predict(X_valid)
    
    # Evaluate the model
    mse = mean_squared_error(y_valid, predictions)
    mse_scores.append(mse)
    
    r_squared = model.score(X_valid, y_valid)
    r_squared_scores.append(r_squared)
    
average_mse = np.mean(mse_scores)
average_r_squared = np.mean(r_squared_scores)

print(f'TRIAL FOR MYC TF 1-MER: Average MSE: {average_mse} Average r_squared: {average_r_squared}')

TRIAL FOR MYC TF 1-MER: Average MSE: 0.00809276745611917 Average r_squared: 0.7750679380615456


In [18]:
myc_sequences = myc_df['binding sequence'].tolist()
myc_values = myc_df['binding affinity'].tolist()

# One-hot encode the sequences
myc_X = np.array([one_hot_encode_2mer(seq) for seq in myc_sequences])
myc_y = np.array(myc_df['binding affinity'])

myc_cv = KFold(n_splits=10, shuffle=True, random_state=1)

In [19]:
# results
mse_scores = []; r_squared_scores = []

# Initialize Linear Regression model
#model = LinearRegression()
#model = Lasso(alpha=0.001)
#model = Ridge(alpha=0.001)
model = ElasticNet(alpha=0.001, l1_ratio=0.5)

# Loop over each fold
for train_index, valid_index in max_cv.split(myc_X):
    # Split data
    X_train, X_valid = myc_X[train_index], myc_X[valid_index]
    y_train, y_valid = myc_y[train_index], myc_y[valid_index]

    # Fit the model
    model.fit(X_train, y_train)
    
    # Predict on valid set
    predictions = model.predict(X_valid)
    
    # Evaluate the model
    mse = mean_squared_error(y_valid, predictions)
    mse_scores.append(mse)
    
    r_squared = model.score(X_valid, y_valid)
    r_squared_scores.append(r_squared)
    
average_mse = np.mean(mse_scores)
average_r_squared = np.mean(r_squared_scores)

print(f'TRIAL FOR MYC TF 2-MER: Average MSE: {average_mse} Average r_squared: {average_r_squared}')

TRIAL FOR MYC TF 2-MER: Average MSE: 0.006143310409632919 Average r_squared: 0.829227326422146


Train your two models: one using the 1mer feature and the other using the 2mer feature. After
training, compare the model’s performance by evaluating their Mean Squared Error (MSE) and
R-squared values. Provide an explanation of your findings. [2pt]

**Summarizing our performance metrics for all of the models we created:**

|           | Mean squared error    | R-squared          |
|-----------|-----------------------|--------------------|
| Max 1-mer | 0.006311913071929501  | 0.78088946154413   |
| Max 2-mer | 0.004261243948972678  | 0.8520113133526607 |
| Mad 1-mer | 0.010255679123879247  | 0.7711734757652452 |
| Mad 2-mer | 0.0067631008309956735 | 0.8491122657585828 |
| Myc 1-mer | 0.00809276745611917   | 0.7750679380615456 |
| Myc 2-mer | 0.006143310409632919  | 0.829227326422146  |

**A general trend we can observe is that the 2-mer approach to one-hot encoding consistently yields enhanced performance across all of the transcription factors which we evaluated. We know that the performance is better because we see lower mean squared error and higher r-squared values for each of the 2-mer models when compared to their 1-mer counterparts. A possible explanation for why the 2-mer models may fare better is that it may be more appropriate to represent the DNA sequences in chunks of multiple nucleotides because transcription factors look for biophysical (i.e. electrostatic) signatures when binding. A protein that binds to DNA does not necessarily solely focus on the sequence of the nucleotides one-by-one, but rather it looks for an ideal pocket of a specific electronegativity, geometry, hydrophobicity, and other biophysical factors. The biophysical features only emerge when we have a system of multiple interacting nucleotides which are close to each other. Therefore, we can only appreciate these features when we break up the sequence into segments of k-mers (where k >= 2). Therefore, a one-hot encoding approach which represents the sequence in 2-mers helps represent the problem at hand in a more appropriate fashion.**

**We observe similar performance across the three transcription factors, meaning that our models are able to relevant and generalizable features of protein binding while not overfitting to just one type/class of protein. Our pipeline can be improved via hyperparameter optimization (mentioned earlier) to find an optimal learning rate (eta) and lambda value. We can also obtain richer HT-SELEX and PBM datasets with a broader class of binding proteins. We could explore more approaches to one-hot encoding as well; perhaps scanning across the sequence for 3-mers, 4-mers, or larger k-mers could be better for extracting biophysical signatures from DNA.**